src/app/CommandHandler.h

```diff
@@ -50,8 +53,34 @@ class CommandHandler : public Command
 
 private:
     friend class TestCommandInteraction;
+    friend class CommandHandlerHandleDeleter;
     CHIP_ERROR SendCommandResponse();
     CHIP_ERROR ProcessCommandDataElement(CommandDataElement::Parser & aCommandElement) override;
+
+    struct CommandHandlerHandleDeleter
+    {
+        void operator()(CommandHandler * apCommandHandler) const;
+    };
+
+    uint32_t mPendingWorkCount = 0;
+
+public:
+    using CommandHandlerAsyncHandle = std::unique_ptr<CommandHandler, CommandHandlerHandleDeleter>;
```

Comment:
One problem with this is that this will have a tendency to run from destructors, which leads to two problems:

1. People don't do a good job of controlling which thread their destructors run on, so it's easy to end up calling into the CHIP stack on the wrong thread.
2. It's probably easier to end up in the destructor here after CHIP stack shutdown than it would be with an explicit (hence visible) function call.

Problem 1 can be ameliorated by having the deleter queue the actual manipulation of the CommandHandler state over to the CHIP event loop.  I'm not sure we can do much about problem 2....

Reply:
The first problem should be discussed in the field of thread model I think, and some other RAII handlers have the same issue.

As for the second issue, I would suggest some kind of sanity checks that will crash (or block?) during shuttingdown.
1. In real apps, since we are shutting down not clean, a crash (or block, and let the watchdog kill the program) makes some sense.
2. In tests, if we cannot shutting down cleanly, then we must have something wrong with the test case or the code, then it should crash.

Reply:
> and some other RAII handlers have the same issue.

Which other RAII handlers do we have that are public API?  The main point is that our public API should be very predictable and/or clearly documented in terms of its behavior.  At least one of those two, ideally both.  In its current state this is neither.

> I would suggest some kind of sanity checks that will crash (or block?) during shuttingdown

I think that would be preferable to memory corruption....

Reply:
> > and some other RAII handlers have the same issue.
> 
> Which other RAII handlers do we have that are public API? The main point is that our public API should be very predictable and/or clearly documented in terms of its behavior. At least one of those two, ideally both. In its current state this is neither.
>

Ah, yes, none of other RAII handlers are designed to be public API. I have updated the code to make the decrease of counter and send response happening on CHIP thread.

> 
> > I would suggest some kind of sanity checks that will crash (or block?) during shuttingdown
> 
> I think that would be preferable to memory corruption....

Just found the current implementation just forced to close all connections. Then one idea comes to my mind that we can use some magic random number to identify the handler with the actual command handler object. If the magic number not match, it will release and do nothing to the command handler, also it can be configured to crash during development.

I'm not sure which one is better, since the calling of Shutdown should be a very rare case on real devices, maybe crash makes more sense.

Reply:
> I'm not sure which one is better, since the calling of Shutdown should be a very rare case on real devices

I don't know about devices, but at least on iOS we call Shutdown on the CHIP stack all the time.  We need to be able to shut the stack down cleanly and safely and restart it.